### PR TITLE
Implement `SkeletonRetarget` and overhaul some animation features

### LINKED ISSUE #43848: Bezier tracks can't fade in/out animations with AnimationTree
**Godot version:**
3,2,2

**OS/device including version:**
Windows 10 , RTX 2080 Super GLS2

**Issue description:**
OneShot node can't fade in/out animations where one animation from animationPlayer is made "normal" and second have animation based on Bezier key's 

**Steps to reproduce:**
repro in zip file

**Minimal reproduction project:**
[oneShot_issue.zip](https://github.com/godotengine/godot/files/5595277/oneShot_issue.zip)

select AnimationTree and test OneShot nodes

### LINKED ISSUE #48526: AnimationTree does not apply Call Method Track for a method in an AudioStreamPlayer script.
**Godot version:**
3.3

**OS/device including version:**
Windows 10 Home, version 20H2, build 19042.928, HP Pavilion Gaming Laptop 16-a0xxx.
nVidia GeForce GTX 1660 Ti Max-Q, driver version 457.63, any backend.

**Issue description:**
Apologies, this is a weird one.

What happens: AnimationTree state machine does not apply method calls to an AudioStreamPlayer script in a Call Method Track in an Animation when another Animation exists on the same AnimationPlayer and which is alphabetically after the current animation and has an Audio Playback Track.

(yes, all of those details are important, see steps to reproduce below)

What I expected: The Call Method Track should get called regardless of other animations' Audio Playback Tracks and alphabetical status.

**Steps to reproduce:**
Gosh, where to begin...

Setup:
1. Create a new Node2D scene with a child Sprite node to animate (optional but useful for debugging).
2. Add AudioStreamPlayer node to the scene with some audio stream imported.
3. Attach a script to the AudioStreamPlayer with a simple method that prints a debug message or something.
    ```gdscript
    extends AudioStreamPlayer
    
    func play_bloop():
	    print("hi")
	    play()
    ```
4. Add an AnimationPlayer node to the scene.
5. Create a simple animation named `default` which translates the Sprite (optional) and which has a Call Method track pointing to the AudioStreamPlayer. Optionally, make this looping.
6. Add a keyframe to the Call Method track which calls the AudioStreamPlayer's method which was added in step 3.
    image
7. Create a second animation named `zzz` which has an Audio Playback track with a keyframe that plays a stream to the AudioStreamPlayer (edit: actually, the keyframe is optional, only the track itself needs to exist).
    image
8. Add an AnimationTree node to the scene.
9. Point the AnimationTree to the AnimationPlayer, mark it as Active, and create a new state machine tree root.
10. In the tree root, add a single node to play the `default` animation.
11. Attach a script to the scene root node which starts the AnimationTree's state machine at the `default` node on ready:
    ```gdscript
    extends Node2D
    
    func _ready():
	    #$AnimationPlayer.play("default")
	    $AnimationTree["parameters/playback"].start("default")
    ```

Actual bug reproduction:
1. Play the scene. Observe that the `default` animation is playing, but the method in the Call Method track is not being called.
2. Do any one of the following:
    - Change the root node script to play the animation via the AnimationPlayer directly, without using the AnimationTree.
    - Remove the Audio Playback track from the `zzz` animation.
    - Rename the `zzz` animation to `aaa`.
3. Play the scene. Observe that the method is now being called as expected.

Additional notes:
- If there are multiple other animations which are similar to `zzz`, they must all be "fixed" in order to get the method calls to function.
- This does not affect a Call Method track which targets a node other than the AudioStreamPlayer.
- *[screaming]*

**Minimal reproduction project:**
[AudioAnimationTreeTest.zip](https://github.com/godotengine/godot/files/6440509/AudioAnimationTreeTest.zip)

### LINKED ISSUE #49431: AnimationTree doesn't blend different track types
### Godot version

3.x, 4.0.dev

### System information

MacOS X, Windows 10

### Issue description

AnimationTree doesn't blend different track types.

### Steps to reproduce

For example, there are two animations, A and B, and one object O.

- In animation A, the object O is moved using the ValueTrack.
- In animation B, the object O is moved using the BezierTrack.

If you blend Animation A and Animation B, they will not blend and only one of them will play.

#### Related issues

- https://github.com/godotengine/godot/issues/48526
- https://github.com/godotengine/godot/issues/43848

This is also the case when a Transform3DTrack is mixed in. Also, some tracks are generated by importing, while others are generated by pressing the key button in the inspector, so it is not easy to unify track types.

### Minimal reproduction project

[track_blends.zip (Godot 4.0 project file)](https://github.com/godotengine/godot/files/6618487/track_blends.zip)

### LINKED ISSUE #54407: Animations not blending properly in AnimationTree. Vertical position of the skeleton is affected.
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0.dev

### System information

Windows 10 - 64 bit

### Issue description

The vertical position of a character seems to have some kind of offset when two different animations are blended together inside an Animation Tree. This is not exclusive to the Blend3 node.

ezgif-1-c62e8a3e6dc8

This behavior is not present in previous versions. I tested the same configuration using Godot 3.4 (beta 6). 
I imported the character model using the .glb file from the TPS Demo. The problem is present in custom characters too.

### Steps to reproduce

The following reproduction project showcases this behavior.

 - Open the "TPS Demo Robot Scene"

### Minimal reproduction project

[gODOT TEST.zip](https://github.com/godotengine/godot/files/7445608/gODOT.TEST.zip)

### LINKED ISSUE #55838: AnimationTree AnimationNodeBlendTree skeleton scaling issue with Blend2 and Blend3
### Godot version

4.0 dev

### System information

Win10, GLES3, RTX 2070 mobile

### Issue description

In 4.0 latest nightly dev build (as of 11 Dec) there is an issue with skeleton scaling when using Blend2 or Blend3 in a AnimationNodeBlendTree in AnimationTree. When blending between 2 animations the skeleton weirdly scales (looks like it's scaling down the root bone in between the blended animations). This issue was not present in Godot 3.4 stable. I used the same gltf import from Blender (skeleton with animations) in 4.0 as in 3.4.

See here: https://www.youtube.com/watch?v=WX0F99OoLQk

### Steps to reproduce

1. Import a gltf character with animations (I used one from Mixamo)
2. Create AnimationTree
3. Select NewAnimationBlendTree
4. Create 2 or 3 Animation nodes
5. Create a Blend2 or a Blend3 node
6. Connect the 2 or 3 animations into the Blend node and to the output
7. Link AnimationTree to AnimationPlayer that have the character animations (inside the imported gltf)
8. Activate the AnimationTree in editor
9. Scrub the Blend back and forth - you will see the skeleton scaling when blending; example here: https://www.youtube.com/watch?v=WX0F99OoLQk

### Minimal reproduction project

Don't have one, just follow the steps above.

### PR DESCRIPTION
**Pull request by TokageItLab**
Implement `SkeletonRetarget` based on reduz's proposal (https://github.com/godotengine/godot-proposals/issues/3379).

This PR includes follows:

- ~#55840~ Superseded by #58863
- #55218
- ~#54205~ Superseded by #57675
- #49411

This requires #55840 because `global_pose` had a serious problem of not being able to get IK and other skeleton modifire transforms. We also needed #54205 to test the `AnimationTree`.

#55218 and #49411 are bug fixes that are not directly related to retarget, but I believe they need to be fixed in the future, and they cause conflicts with this PR, so I included them in this PR to resolve the conflicts in advance.

Fixed #43848. Fixed #48526. Fixed #49431. Fixed #54407. Fixed #55838.

---

### SkeletonRetarget (Node)

Transfers the transform between the two skeletons. I implemented three types of resources for retarget. It corresponds to the "Retargeter" resource that reduz mentioned in the proposal, which is separated into several parts to make it easier to reuse.

- RetargetProfile / RetargetRichProfile
- RetargetBoneMap
- RetargetBoneOption

`RetargetProfile` provides the intermediate bones. The source skeleton and target skeleton bones are mapped to each other through intermediate bones using `RetargetBoneMap`. The transfer value is calculated differently by `RetargetMode`. The default mode is `RetargetMode::RETARGET_MODE_GLOBAL`. You can set it for each intermediate bone using `RetargetBoneOption`.

### RetargetProfile (Resource)

This resource is special since it is only used in the editor. It does not have any effect on the retargeting calculation.

The `RetargetBoneMap` is needed for each skeleton with a different bone name, but the keys must be the same. Therefore, to make it easier to generate the same keys in those, `RetargetProfile` lists the keys as intermediate bones.

You need a `RetargetProfile` when you register a key to `RetargetBoneMap` or `RetargetBoneOption` in the editor. However, once you register the keys in them, you can discard the `RetargetProfile` if you don't need it. Also, you can register keys in them with the script always.

`RetargetRichProfile` defines textures, coordinates, and groups in addition to the intermediate bones. It provides a GUI with the buttons placed on the silhouette image to register the keys to `RetargetBoneMap` and `RetargetBoneOption`.

If users have defined their own RetargetProfile, it is recommended that they also include this profile when sharing retarget animations to make convert easy between users.

I'm still undecided on how to implement the presets that Godot has. Should I just distribute `gd_humanoid.tres` in the Asset library, or should I provide a hard-coded uneditable `GDHumanoidProfile` class that inherits from `RetargetRichProfile`? A similar implementation used to exist, `default_env.tres`, but that has been replaced by an internal implementation.

### RetargetBoneMap (Resource)

A map of skeleton bone names with intermediate bone names as key. To register a key in the editor, a `RetargetProfile` is required. In the inspector, maps are organized by `RetargetProfile`, but if there is no `RetargetProfile`, they are listed in the unprofiled bones section.

### RetargetBoneOption (Resource)

Set the options for each intermediate bone. It has `RetargetMode` as one of the options. The properties that should be set for each intermediate bone and do not need to be set for each skeleton, will be put together in `RetargetBoneOption` instead of `RetargetBoneMap`.

#### RetargetMode::RETARGET_MODE_GLOBAL

Transfers the global transforms in the model space relative to the bone rest.

If the models are in the same stance, the animations will basically transfer well. However, if the models are in different stances, such as T-stance and A-stance, the transfer will be made looks strange.

However, the retargeter should not involved in that. Potentially, it make sense that an add-on such as T-poser makes match the stance between them.

BTW, the bones do not allow for shearing, so it will be orthogonalized implicitly if the animation includes scaling.

#### RetargetMode::RETARGET_MODE_LOCAL

Transfers the local transforms relative to the bone rest (same as the default animation format in Godot 3).

https://user-images.githubusercontent.com/61938263/149949487-c399962e-6383-4766-b17c-aa9128575614.mp4

For example, for bones that are slanted even at T-stance, such as fingers, the joints may rotate in odd directions. In such a case, it is recommended to use `RETARGET_MODE_LOCAL` with a restriction in the bone rest axis such as "+X axis rotation to bend".

#### RetargetMode::RETARGET_MODE_ABSOLUTE

Transfers the local transforms relative to the initial value of `Transform` (same as the default animation format in Godot 4).

If the bone rest does not match completely, it will look bad. Possibly, it could be useful for transferring animations between different stances, such as T-stance and A-stance, if the Rest rules match, but it is basically not recommended to use it for transferring the transforms between different models.

---

## AnimationPlayer

`SkeletonRetarget` transfers the transforms from the source skeleton to the target skeleton, but this means there are always two skeleton in the scene, which can make for worse performance in some cases.

Therefore, I implemented the properties for retarget in the `AnimationPlayer` to make it easier to manage the animation and replace the target model.

Also, as a variation of `Pos/Rot/Scl3DTrack`, implemented retarget track which has `RetargetMode` (actually it is just `Pos/Rot/Scl3DTrack` which has special path like `:Hips`, only subpath).

To extract the retarget tracks, it need a source skeleton and a `RetargetBoneMap`. In other words, once the retarget tracks are extracted, the source skeleton is no longer needed.

You can create the retarget tracks from the options.

The retarget track will transform to the skeleton assigned to the `AnimationPlayer`.

https://user-images.githubusercontent.com/61938263/149949721-dffae5e9-4815-4b4e-b4ea-eee401c2d6d9.mp4

This will make it easier to manage animation resources and distribute retargeted animation packages grouped by `RetargetProfile` in asset libraries, etc.

BTW, I'm worry about  that should it allow multiple retarget skeletons to be assigned to an `AnimationPlayer`, in case one animation contains multiple skeletons? In my opinion, that can be solved by splitting the animation into multiple animations and using `AnimationPlaybackTrack`.

**Note:**
For rotation, the global retarget track will calculate the multiplication `Quaternion` 3 extra times per bone, and the local retarget track will calculate the multiplication 1 extra time. As a reference, for a single model with 54 bones, using or not using the global retarget track made a difference of about 0.2ms in the process.

I think this is not enough of a performance hit to be a problem for several models, but I would recommend matching the axes in the DCC software and using local retarget if possible (the delay will be 1/3). Of course, it goes without saying that if you have hundreds of enemy models, you shouldn't use global/local retarget in runtime in the first place; You can restore retarget tracks into them.

---

### To adjust position value between the models have different height

The position values are transferred regardless of the model height. For example, transferring the root motion from a short model to a tall model will cause a slip. Position values are may used in weapon throwing animations. For toe and knee stands, we may need to specify specific bones in the animation. For example, a toe stand requires the length of the Foot-Toes. Kneeling requires a length of LowerLeg-Foot. And its length varies from model to model, both require a Hips bone to move.

For now, I think the only problem is the root motion, but I don't know if I should solve it in the retarget process. It may be solved by implementing an AnimationTree with a `root_scale` property. Or, maybe we need to store RootDistance (ModelScale / HipsHeight) as a single float value in RetargetBoneMap in the future. So, I make it is out of the scope of this PR.

### For compressed tracks

I'm a little concerned about the conversion of compressed tracks, since it convert all keys when converting to Retarget tracks, but for compressed tracks it may need to do the same for all byte arrays. This will be worked out after the implementations around of the compressed/stream track is finalized.

### Retarget track renamer 

We may want a utility to match the names between extracted animations with different profiles, it can be added later.

### Adjusting the transform value when the target model has fewer bones than the source model

For example, if you are transferring motion from a model with four spine bones to a model with three spine bones, you will need to add the deformations of the removed bones to one of the three in order for the looks to match.

I experimented with this a bit, but aborted because it seems that we would need to change the cache format of the animation system to support this.

One solution that can be implemented quickly is to add a virtual bone of zero length to the parent or child. This helper method can be added later.

[skeleton_retarget_sample_project.zip](https://github.com/godotengine/godot/files/8200580/skeleton_retarget_sample_project.zip)

### CODE REVIEW

**Comment by TokageItLab** [comment]
@Calinou It's 3D only, do I need to rename the class to SkeletonRetarget3D in addition to putting it in the Skeleton document?

**Comment by Calinou** [comment]
> @Calinou It's 3D only, do I need to rename the class to SkeletonRetarget3D in addition to putting the note in the SkeletonRetarget document (class reference)?

I'm not sure if the class should be renamed – I'll let others give their opinion on this.

**Comment by TokageItLab** [comment]
Updated the documentation for now, clarifying that SkeletonRetarget is for Skeleton3D in the class reference.

I don't have much knowledge about Skeleton2D. If it has the concept of rest/pose, then I think a similar implementation is possible. However, I think 2D assets don't have much concept of stance, unlike 3D assets, so I guess there is not much needs for it... I also wait for others' opinions.

**Comment by TokageItLab** [comment]
The implementation is large due to the new EditorPlugin for editing the map, so I'm rebasing it quite often to check for hidden conflicts, but it's always ready for review.

**Comment by Ansraer** [comment]
Huh, I was just looking for a plugin that did this. Thanks for writing a PR!
Will build this PR in the next few days and test things myself, but from glancing at the description I am a bit concerned about the position problems when source and target have different heights. Feels like something that really needs to be fixed, since root motion is quite common afaik.

The best solutions my sleep-deprived brain can come up with right now are either 
* scale all movement based on the source/target height difference (maybe allow users to enable/disable this based on the animation that is currently playing?)
* ensure that after the animation is applied to the target the distance between the target's lowest vert and the ground is not bigger than the distance between the source's lowest vert of and the ground at that frame.

**Comment by TokageItLab** [comment]
Right, I know we need it and I'm considering with @fire, @SaracenOne and @lyuma :

- Is a single float value such as HipsHeight sufficient?
- Which resources and nodes should store the value?
- Should that apply not only to root motion, but also to Hips and other bone movement?

**Comment by fire** [comment]
Working with @TokageItLab to pull out 55840 so there's less to review.

**Comment by TokageItLab** [comment]
Fixed type and resolved conflicts. Still I keep rebasing once a day, but it is always ready for review.

**Comment by TokageItLab** [comment]
Made adopt to #57675 instead of #54205.

File: scene/resources/animation.h [hunk 1]
```diff
@@ -77,6 +77,12 @@ class Animation : public Resource {
 		HANDLE_MODE_BALANCED,
 	};
 
+	enum RetargetMode {
```
Comment (reduz) [inline scene/resources/animation.h:80]: It does not seem correct that retarget information has to exist in Animation. What is the rationale behind this?

Comment (TokageItLab) [reply scene/resources/animation.h:80]: @reduz It is obviously redundant to always have two Skeletons and need a SkeletonRetarget node when using retargeted animations for several models.

If you don't include that mode in your animation, we can store the Rest itself in the animation data, and for example, for rotation, you will need to multiply the quaternion 6 times during playback.

However, with the method I implemented, we can split the process into two parts: extract and apply. When playback animation, all that is required is 3 multiplications in the apply process.

**Comment by TokageItLab** [comment]
We're struggling with @reduz on whether to record `retarget_mode` in `Animation::Track` or `Skeleton3D::Bone`.

The fundamental difference between my approach and reduz's is the point that "animation requires or does not require a Skeleton from which the animation is generated".

---

The premise is that animation data can have information about the original Skeleton through pre-calculation. However, this is destructive and the information of the original Skeleton cannot be retrieved from the animation.

For example, let's say the bone rest value is `1` and the animation value is `3`.

```
3 - 1 = 2
```

In this case, a value of `2` is required for retargeting. Therefore, the animation only needs a value of `2` to be ready for retargeting.

What is being debated is where to record the "information as to whether the value is a pre-calculated value or not".

I believe it should be stored in the same place as the value. In other words, `retarget_mode` should be recorded in `Animation::Track`. I also believe that the original Skeleton is no longer needed at this point.

Reduz thinks it should be selected as a preset when importing the model and store the `retarget_mode` in `Skeleton3D::Bone`. And, he believes that the animation should be pre-calculated for performance, but the original Skeleton should continue to be stored somewhere.

---

I consider the reduz method to be problematic in two main points.

**1) If a preset is changed midstream, all animation values must be recalculated**

Also, the original skeleton which is source of imported animations is required for recalculation.

It means that they will be broken if the original skeleton has been removed and does not exist, or if there are duplicates or newly created animation that are not tied to the original skeleton.

**2) It is to be impossible to blend LocalAnimation and GlobalAnimation**

This is clear if we assume that presets are to be used, but if you store the `retarget_mode` in `Skeleton3D::Bone` without using a preset, they will not blend well.

LocalAnimation and GlobalAnimation cannot be blended directly because they are recorded in different coordinate spaces by pre-calculation. In order to blend them, their values must be restored to the target skeleton bone's coordinate space, but to do so, it is necessary to know whether the animation was "pre-calculated as LocalAnimation" or "pre-calculated as GlobalAnimation".

There is no need to think too hard, it is obvious that if you have an animation extracted as GlobalAnimation, its playback will be broken if LocalAnimation mode is applied to the target Skeleton.

We are discussing hard about this case of blending LocalAnimation and GlobalAnimation because reduz claims that it is a very niche case.

---

For example, I have Model A and Model B, with rest optimized for my project. These models have different rest values, but the **rules** for the joints match, such as +X rotation bends, and parent to child is Y-axis and roll.

The basic animation is created by Model A and shared with Model B as LocalAnimation. 

It is important to understand that LocalAnimation is less likely to cause bone breaks if the models match the joint rules. This is very useful in cases where different models have different orientations, such as the thumb. If you have struggled with finger control in Unity models, this is probably easy to understand.

However, I want to add more animations, so I want to import animations from Model M, where Rest rules do not match.

**If `retarget_mode` is recorded in `Animation::Track`:**

the animation of Model M can be imported as `GlobalAnimation` and applied to Model A or Model B, moreover it can be blended with LocalAnimation.

**If `retarget_mode` is recorded in `Skeleton3D::Bone`:**

there are two ways.

Import the animation of model M as GlobalAnimation. However, re-import the animation of model A and model B as GlobalAnimation instead of LocalAnimation. At this time, the advantage of LocalAnimation's bone breaking resistance is lost, and the playback animation of the fingers will be dirty.

Alternatively, the rest rule of Model M could be matched to Model A and Model B by somehow, but the process is complicated.

---

TL;DR


- I believe that having `retarget_mode` in `Animation::Track` would make it easier to share only animations among users and projects
	- Reduz believes that .tres(animation) should not be allowed to be used in separate from Skeleton in the first place, so `retarget_mode` should be stored into the `Skeleton3D::Bone`
- I use LocalAnimation because I need to care about finger animation in certain animations. At the same time, I use GlobalAnimation for other animations where I don't need to care about finger animation. Then I blend them together
	- Reduz says that "this case is that you are the only one who will use it"


We need others' opinions as it is difficult for us to resolve the discussion on our behalf.

Please give us your opinion if possible: @fire @lyuma @SaracenOne @AndreaCatania @TwistedTwigleg @nonunknown 

If any other developers are interested in character animation or retargeting, please join the discussion.

**Comment by Zireael07** [comment]
Hm. Looking at my character models, who have different bone names and rest poses despite coming from the same author and pack, I think I would mostly use GLOBAL mode. However, for some things, e.g bending arms to wave, or legs to sit down, I think LOCAL would be more intuitive (just bend this by 90 degrees, this other by 90 degrees, done).

So yes, +1 to something that allows mixing GLOBAL and LOCAL.

**Comment by lyuma** [comment]
In short, having played with this, and talked this over with others, I'm trying to boil down some of the things I observed.

First of all, Local (rest) and Absolute modes are not good enough for most of the common retargeting cases. As a test,. I transferred a local animation from one character to an Xbot model, and the animation was broken. See here: left is Global (GLO) which works well. Right is Local (LOC) which has arms and body backwards (image below).

So the question is: when is Global not good enough? Well, it is not well suited for finger bones. By its very nature, it does not know how finger bones work, since it is working on the "silhouette" of the model.

This means none of these retargeting modes are good enough on their own. So what can we do about it?

- Perhaps it is possible to create an importer which twists fingers in the Bone Rest matrix into a configuration suitable for Global retargeting? I can't say whether or not this is possible, but if this could be solved, then maybe we can use Global mode for everything?
- It could be declared that finger bones shall be local, and all other bones shall be global, and everyone making assets should agree on finger rotations. This might not always be possible, but this is my interpretation of reduz's idea of storing the PlaybackMode in Node B in Tokage's diagram.
- Or it could be left up to each animation artist / importer which to use, and then you're going to have a mixture of modes. If so, I see two ways to deal with it:
1. The animations are stored in global along with skeleton and converted using the source skeleton. This sounds ok in principle if some conversion tool could be made, but I would be afraid to deal with converting animations on the fly at runtime.
2. the animations must point to the RetargetOption they were generated with (or a per-track retarget mode), so that these differently sourced animations can be mixed.

from what I can understand, "2" is what Tokage is proposing. I don't see many other solutions to this. Fundamentally, if it is the case that animations made from different models are needed to be mixed, I don't see a viable solution other than the last one.

**Comment by TokageItLab** [comment]
@lyuma 
> First of all, Local (rest) and Absolute modes are not good enough for most of the common retargeting cases.

Yes, it has been proven in Godot 3.x that it is not sufficient for retargeting.

For the time being, **GlobalAnimation should be the default to avoid confusion for light users**. However, for advance users, for parts with delicate joints such as fingers, it is more appropriate to use LocalAnimation (RestAnimation) with matching RestRule.

MMD (MikuMikuDance), a dance creation software, basically only performs GlobalAnimation, but there are plug-ins that adjust fingers' Rest for editing, which proves that there is a demand for this method.

And for AbsoluteAnimation, the only use case I can think of for transferring AbsoluteAnimation to another model is when you want to animate an IK target as a bone which is child of the root, regardless of the model's figure. It does not seem to be completely useless.

---
> Perhaps it is possible to create an importer which twists fingers in the Bone Rest matrix into a configuration suitable for Global retargeting? I can't say whether or not this is possible, but if this could be solved, then maybe we can use Global mode for everything?

This is in principle the same as making an A-pose into a T-pose; doing it automatically is not sufficient. There must always be manual confirmation, or the joint may bend in a strange direction.

For reference, some MMD plug-ins will align the RestRule to some degree, but correct looks requires manual adjustment finally.

---
Even if the rest rule matches, GlobalAnimation should be used for Hips and Spine.

The reason is that they do not require as much rigorous deformation because the joints are not clearly visible, and because they are closer to the root, there are more errors that depend on the model.

For example, when sharing "an animation of a horizontal waist twist created with a model that does not have an S-shaped waist", using LocalAnimation may cause the arms to turn at an angle on a model with an S-shaped waist.

In the same way, Neck-Head and Shoulders should basically use GlobalAnimation, but in some cases such angled rotation may be desired, so blending with LocalAnimation should be possible.

**Comment by TokageItLab** [comment]
I wanted to keep RetargetMode in the AnimationLibrary implemented in #59980 and be able to blend different RetargetModes, but @reduz rejected that.

---

Part of the discussion:

Tokage:
> Let me explain the use case again.

> In an extreme use case, a normal human and a robot whose knee Rest is facing 180 degrees in the opposite direction, want to share an animation of bending their knees. In this case, the human and robot have the same Rest-Rule, so RestAnimation can bend the knee in the correct direction. However, with GlobalAnimation, the knee will bend in the wrong direction.

> A 180-degree error is fatal. In fact, if you have several different human models, the finger angles can be off by as much as 15 degrees from one model to the next. This is because fingers do not have a proper format like T-pose. Therefore, it is necessary to use RestAnimation for fingers and to match Rest-Rule as much as possible.

> The problem is that when the Rest-Rules share animations of different models, the animations are broken unless they are GlobalAnimation.

> Or the Rest-Rules must be matched. This should be done for everything, but this is very costly. Your proposal requires this.

> Therefore, the best way to maximize performance is to use RestAnimation for animations like dancing, which requires a large finger on the screen, and use GlobalAnimation from different assets for situations like moving a field, which requires many animations. I believe that this is right way.
If the number of models is 10 and the number of animations is 50, my proposal only requires 10 Rest-Rules to be matched, but your proposal requires it for 60
Incidentally, the Rest-Rule matching must be manually done. Assistance by script is possible, but visual confirmation is always required

reduz:
> I understand as much, but this sounds like its something you will align per model, not per animation

Tokage:
> Well, in reality, since 50 animations are not all on different models, the number of animation source models will be about 10, but it is still a hassle to make 20 adjustments

reduz:
> I understand where you are coming from, but I insist again that this is now the kind of situation you will find when doing actual game development.
It sounds to me that this is something that may commonly happen when you share MMD animations, and it makes sense, but here as a developer you have more control of what you are getting into your game.

Tokage:
> For example, if we are going to allow users to upload custom animations, such as VRchat, we will be forcing them to follow the only one Rest rules

reduz:
> I understand this, but I feel this is not what we are aiming for in Godot, this is a very specific use case for one specific type of game. I think you will have more luck here writing your own animation format and blending logic for a game like this
You have to think that there are a zillion types of games made with Godot, if we would make Godot work exactly how every single of them needs, Godot would be full of features that nobody would use.

Tokage:
> So you are saying that light users don't use RestAnimation and advanced users who use RestAnimation should suffer from Rest matching on Godot

reduz:
> I am not sure if I would consider those advanced users, its a very specific type of game that needs this. To begin with, games with some significant (not even high) budget just create their own animations.
So for the most part, its a very very specific type of game that needs something like this.

Tokage:
> Indie game developers also create their own animations, but it is impossible to create all the animations, so we need to blend the asset animations with your own animations

reduz:
> I do agree with this, but from there to trying to implement a very complex retargeting system that only you need so far there is a long distance.
---

Also, this PR is starting to get old, so I am abandoning it for now. I think the GUI for mapping and the functions to extract transforms for retargeting are salvageable for future implementations of Retarget.

I think more user voices are needed on this. Unfortunately, I feel that not many people who use Godot are familiar with the 3D character workflow. And Godot's Animation team alone cannot make a fair enough decision.

If anyone interested in 3D character workflow and Retarget read the log above, it would help us a bit if you could leave a comment. Or we would like to have a specific discussion on [Contributors Chat](https://godotengine.org/article/new-communication-platform-godot-contributors).

**Comment by Zireael07** [comment]
As a prospective user (I have several character assets, one of them is not from the packs the others are, and even the others don't share skeletons even though they share the author) I would like even a basic retargeting system to make it in. More advanced stuff like what you seem to be discussing with reduz above can go in in a further PR IF it turns out to be needed.

**Comment by JoanPotatoes2021** [comment]
Here's how I would expect animation retargetting to work, my opinion in a grain of salt, as an artist and game developer I can work on my own animations for anything I need, however what I can't do is to customize/mass produce animations to each different character I want with different proportions, that would take too much work, unreal show a use-case with it's animation retargeting that describes exactly what I want as a developer, to re-use animations on a somewhat similar or identical humanoid skeleton with different proportions. Here's the link to unreal doc:

https://docs.unrealengine.com/4.26/en-US/AnimatingObjects/SkeletalMeshAnimation/AnimationRetargeting/

My idea of skeleton retargeting is not to be able to gain animation assets from different skeletons from some rig on the internet, but to increase the possibilities with my animations to adopt new proportions. Here are some photos of what I described gathered from the unreal docs.







And here's a definition from unity's "Retargeting of Humanoid animations" section:



I could only dream of a animationblendtree node filtered on which animation I want to be retargetted on which bones.
